### PR TITLE
Token-based access for the Usage API

### PR DESCRIPTION
**Pull request by borisroman**

### CODE REVIEW

File: cosmic-usage-api/src/main/java/com/github/missioncriticalcloud/cosmic/api/usage/exceptions/UnAuthorizedException.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
+package com.github.missioncriticalcloud.cosmic.api.usage.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.UNAUTHORIZED)
+public class UnAuthorizedException extends RuntimeException {
```
Comment (eliasgomes) [inline cosmic-usage-api/src/main/java/com/github/missioncriticalcloud/cosmic/api/usage/exceptions/UnAuthorizedException.java]: _Unauthorized_ is a single word.

File: cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package com.github.missioncriticalcloud.cosmic.usage.core.services;
+
+import java.io.IOException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.github.missioncriticalcloud.cosmic.usage.core.model.DomainToken;
+
+public interface TokenService {
+
+    String encryptDomainToken(final String domain) throws JsonProcessingException;
```
Comment (eliasgomes) [inline cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java]: I'd rename this method to just `String encrypt(DomanToken)` and I'd wrap vendor-specific exceptions like `JsonProcessingException` with something like `TokenException`

File: cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+    String encryptDomainToken(final String domain, final long ttl) throws JsonProcessingException;
```
Comment (eliasgomes) [inline cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java]: I'd remove this method as the `TTL` should be specified directly in the `DomainToken`, in my opinion. Default `TTL` value could turn into a constant in `DomainToken`.

File: cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+
+    DomainToken decryptDomainToken(final String token) throws IOException;
```
Comment (eliasgomes) [inline cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java]: I'd rename this method to `DomainToken decrypt(String)` and wrap the exception

File: cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+    boolean verifyDomainToken(final String token, final String domain) throws IOException;
```
Comment (eliasgomes) [inline cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/TokenService.java]: I'd rename this method to just `isValid`

File: cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/impl/TokenServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+@Service
+public class TokenServiceImpl implements TokenService {
+
+    private final TextEncryptor encryptor;
+    private final ObjectMapper objectMapper;
+
+    private final long ttl;
+
+    private final DateTimeZone dateTimeZone = DateTimeZone.UTC;
```
Comment (eliasgomes) [inline cosmic-usage-core/src/main/java/com/github/missioncriticalcloud/cosmic/usage/core/services/impl/TokenServiceImpl.java]: In my opinion the environment should be responsible for defining the timezone, not the app.

File: cosmic-usage-api/src/main/java/com/github/missioncriticalcloud/cosmic/api/usage/controllers/MainController.java [hunk 7]
```diff
@@ -51,15 +60,27 @@ public Report general(
...
     ) {
+        authenticate(token, path);
+
         final Report report = usageCalculator.calculate(from, to, path, unit, true);
         SortingUtils.sort(report, sortBy, sortOrder);
 
         return report;
     }
 
+    private void authenticate(final String domainToken, final String domainPath) {
```
Comment (eliasgomes) [inline cosmic-usage-api/src/main/java/com/github/missioncriticalcloud/cosmic/api/usage/controllers/MainController.java]: I'd move this method to `TokenService` and call it `validate`.

**Comment by borisroman** [comment]
go build

**Pull request merged**